src/library/scala/util/control/Using.scala

```diff
@@ -0,0 +1,160 @@
+package scala.util.control
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import scala.util.Try
+
+// TODO: scaladoc with example
+/** A utility for performing automatic resource management.
+  *
+  * @example
+  * {{{
+  * for {
+  *   r1 <- Using(resource1)
+  *   r2 <- Using(resource2)
+  *   r3 <- Using(resource3)
+  *   r4 <- Using(resource4)
+  * } yield {
+  *   // use your resources here
+  *   r1.lines ++ r2.lines ++ r3.lines ++ r4.lines
+  * }
+  * }}}
+  * @example
+  * {{{
+  * val lines: Try[List[String]] = Using(resource1) { r1 =>
+  *   r1.lines.toList
+  * }
+  * }}}
+  */
+final class Using[R: Using.Resource](resource: => R) {
+  private[this] val used = new AtomicBoolean(false)
+
+  @throws[IllegalStateException]("if the resource has already been used")
+  @inline def apply[A](f: R => A): Try[A] = map(f)
+
+  @throws[IllegalStateException]("if the resource has already been used")
+  def map[A](f: R => A): Try[A] = Try { useWith(f) }
+
+  @throws[IllegalStateException]("if the resource has already been used")
+  def flatMap[A](f: R => Try[A]): Try[A] = Try { useWith(f) }.flatten
+
+  @throws[IllegalStateException]("if the resource has already been used")
+  @inline private[this] def useWith[A](f: R => A): A =
+    if (used.getAndSet(true)) throw new IllegalStateException("resource has already been used")
+    else Using.using(resource)(f)
+}
+
+object Using {
+  /** Creates a `Using` from the given resource. */
+  def apply[R: Resource](resource: => R): Using[R] = new Using(resource)
+
+  /** Performs an operation using a resource, and then closes the resource,
+    * even if the operation throws an exception. This method behaves similarly
+    * to Java's try-with-resources.
+    *
+    * It is highly recommended to use the `Using` construct, which safely wraps resource
+    * usage and management in a `Try`.
+    *
+    * If both the operation and closing the resource throw exceptions, the one thrown
+    * when closing the resource is
+    * [[java.lang.Throwable.addSuppressed(Throwable) added as a suppressed exception]]
+    * to the one thrown by the operation, ''unless'' the exception thrown when closing
+    * the resource is [[scala.util.control.NonFatal fatal]], and the one thrown by the
+    * operation is not. In that case, the exception thrown by the operation is added
+    * as a suppressed exception to the one thrown when closing the resource. If an
+    * exception is a [[scala.util.control.ControlThrowable ControlThrowable]], no
+    * exception will be added to it as a suppressed exception.
+    *
+    * @param resource the resource
+    * @param body     the operation to perform with the resource
+    * @tparam R the type of the resource
+    * @tparam A the return type of the operation
+    * @return the result of the operation, if neither the operation nor
+    *         closing the resource throws
+    */
+  def using[R: Resource, A](resource: R)(body: R => A): A = {
```

Comment:
I don't think `Using.using` is good nomenclature.  Maybe `Using.direct` or `Using.unsafe` or `Using.unwrapped`?

Reply:
my only concern is that I'd prefer a method name that, if imported statically, still makes sense.

e.g. `Using.direct(file.open)` makes sense, but `direct(file.open)` doesn't really

Reply:
`withResource`

Reply:
or another possibility: `tryWith`

Reply:
(for comparison)

```scala
Using.withResource(new FileInputStream("foo.txt")) { _ => () }
withResource(new FileInputStream("foo.txt")) { _ => () }

Using.resource(new FileInputStream("foo.txt")) { _ => () }
resource(new FileInputStream("foo.txt")) { _ => () }

Using.tryWithResource(new FileInputStream("foo.txt")) { _ => () }
tryWithResource(new FileInputStream("foo.txt")) { _ => () }

Using.tryWith(new FileInputStream("foo.txt")) { _ => () }
tryWith(new FileInputStream("foo.txt")) { _ => () }
```